diff --git a/src/terminal/stream_terminal.zig b/src/terminal/stream_terminal.zig
index e3d42c51e..6f118f2d1 100644
--- a/src/terminal/stream_terminal.zig
+++ b/src/terminal/stream_terminal.zig
@@ -659,9 +659,47 @@ pub const Handler = struct {
                     mask.* = .initEmpty();
                 },
 
-                .query,
-                .reset_special,
-                => {},
+                .query => |kind| report: {
+                    const write_pty = self.effects.write_pty orelse break :report;
+                    const value = switch (kind) {
+                        .palette => |i| self.terminal.colors.palette.current[i],
+                        .dynamic => |dynamic| switch (dynamic) {
+                            .foreground => self.terminal.colors.foreground.get() orelse break :report,
+                            .background => self.terminal.colors.background.get() orelse break :report,
+                            .cursor => self.terminal.colors.cursor.get() orelse
+                                self.terminal.colors.foreground.get() orelse break :report,
+                            else => break :report,
+                        },
+                        .special => break :report,
+                    };
+
+                    var buf: [48]u8 = undefined;
+                    const response = switch (kind) {
+                        .palette => |i| try std.fmt.bufPrintZ(
+                            &buf,
+                            "\x1b]4;{d};rgb:{x:0>4}/{x:0>4}/{x:0>4}\x1b\\",
+                            .{
+                                i,
+                                @as(u16, value.r) * 257,
+                                @as(u16, value.g) * 257,
+                                @as(u16, value.b) * 257,
+                            },
+                        ),
+                        .dynamic => |dynamic| try std.fmt.bufPrintZ(
+                            &buf,
+                            "\x1b]{d};rgb:{x:0>4}/{x:0>4}/{x:0>4}\x1b\\",
+                            .{
+                                @intFromEnum(dynamic),
+                                @as(u16, value.r) * 257,
+                                @as(u16, value.g) * 257,
+                                @as(u16, value.b) * 257,
+                            },
+                        ),
+                        .special => unreachable,
+                    };
+                    write_pty(self, response);
+                },
+                .reset_special => {},
             }
         }
     }
